kube-controllers/pkg/controllers/pod/pod_controller.go

```diff
@@ -321,7 +321,9 @@ func (c *podController) syncToCalico(key string) error {
 		new := wepData.(converter.WorkloadEndpointData)
 		if !reflect.DeepEqual(old, new) {
 			// The relevant wep data has changed - update the wep and write it to the datastore.
+			c.workloadEndpointCache.RLock()
 			log.Infof("Writing endpoint %s with updated data %#v to Calico datastore", key, new)
```

Comment:
If we need a lock to print out the `new` variable then the `reflect.DeepEqual()` on line 324 is also suspicious.  Looks like this whole are of code needs review for concurrency issues.

Reply:
I think we should lock the `key` variable, and I noticed that elsewhere in this function, whenever this variable is read, the `RLock` function is called.

Reply:
I don't think the `key` variable is the problem here

The other lock instances aren't protecting `key` - they're protecting the `workloadEndpointCache` which is accessed from multiple routines.

Reply:
@cyclinder @fasaxc I suspect that the "concurrent map access" issue we're seeing here is because "new" is an instance of `WorkloadEndpointData` which has a map as a field: https://github.com/projectcalico/calico/blob/77ecdcda90c0a8a38ae247067169523484c6c8e3/kube-controllers/pkg/converter/pod_converter.go#L37

Likely we're hitting this line at the same time that we are updating that Label map elsewhere? 

If so, I think we definitely need a lock around the `DeepEqual` above as well.

This gets a bit tricky, because we need to make sure we're locking all of the relevant places that this map may be updated. The `workloadEndpointCache` lock is probably one of them, but there is also the `resourceCache` to think about.

It might make sense to make a copy of the labels when building the `WorkloadEndpointData` so that updates to the pod aren't acting on the same map as the wepData here? Would that allow us to simplify this? https://github.com/projectcalico/calico/blob/77ecdcda90c0a8a38ae247067169523484c6c8e3/kube-controllers/pkg/converter/pod_converter.go#L52-L64

Reply:
I've lost the thread a bit, but  any chance this fix would help: https://github.com/projectcalico/calico/pull/9039.  If we got the WEPs above from the converter we'll have copied the labels already after my fix.

The bug I was fixing was that we were mutating the map that we got from the caching informer, but it also mutates the map when refreshing the cache.

Reply:
Yeah, think there's a good chance that fixed this already :+1: